### PR TITLE
neovim: add a checkPhase in the wrapper

### PR DESCRIPTION
**Pull request by teto**
add a checkPhase (if doCheck == true, disabled by default) that tries to start neovim and fails in case of errors. It allows to catch small misconfiugurations before switching to a new config.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by khaneliman** [comment]
Seems to crash nixpkgs-review
image

File: pkgs/applications/editors/neovim/tests/default.nix [hunk 1]
```diff
@@ -55,9 +55,11 @@ let
   nvim-with-luasnip = wrapNeovim2 "-with-luasnip" (makeNeovimConfig {
     plugins = [ {
         plugin = vimPlugins.luasnip;
-
       }
     ];
+    doCheck = true;
+    # build should fail when this gets uncommented
+    # luaRcContent = "toto";
```
Comment (GaetanLepage) [inline pkgs/applications/editors/neovim/tests/default.nix:62]: Is this supposed to stay ? I see how it can be an example.

File: pkgs/applications/editors/neovim/wrapper.nix [hunk 2]
```diff
@@ -147,7 +148,7 @@ let
 
     pname = "neovim";
     version = lib.getVersion neovim-unwrapped;
-  in {
+  in builtins.removeAttrs attrs ["extraPython3Packages" "extraLuaPackages" ] // {
```
Comment (GaetanLepage) [inline pkgs/applications/editors/neovim/wrapper.nix:151]: Is that related to enabling the check phase ?

File: pkgs/applications/editors/neovim/tests/default.nix (hunk 1 above)
Comment (teto) [reply pkgs/applications/editors/neovim/tests/default.nix:62]: I was not sure how to test for a build failure. So I added this to be able to check. Happy to remove it or you have another suggestion

File: pkgs/applications/editors/neovim/wrapper.nix (hunk 2 above)
Comment (teto) [reply pkgs/applications/editors/neovim/wrapper.nix:151]: instead of adding `inherit doCheck` I thought I could remove all the `inherit withPython, etc` by directly reusing the removeAttrs. Except I haven't removed the other `withXX` so it's a bit redundant :/ Had to remove everything that is a function from attrs since nix cant serialize it

**Comment by khaneliman** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 352727`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>gnvim</li>
    <li>lunarvim</li>
    <li>neovim-qt</li>
    <li>nvimpager</li>
    <li>vim-startuptime</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>neovide</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>gnvim</li>
    <li>lunarvim</li>
    <li>neovim-qt</li>
    <li>nvimpager</li>
    <li>vim-startuptime</li>
  </ul>
</details>

File: pkgs/applications/editors/neovim/tests/default.nix (hunk 1 above)
Comment (MattSturgeon) [reply pkgs/applications/editors/neovim/tests/default.nix:62]: > I was not sure how to test for a build failure.

You can use [`pkgs.testers.testBuildFailure`](https://nixos.org/manual/nixpkgs/unstable/#tester-testBuildFailure) to make a test that passes when the derivation you pass it fails.

File: pkgs/applications/editors/neovim/wrapper.nix (hunk 2 above)
Comment (GaetanLepage) [reply pkgs/applications/editors/neovim/wrapper.nix:151]: Ok I see... Feels a bit hacky no ?

Comment (teto) [reply pkgs/applications/editors/neovim/wrapper.nix]: that's how most builders work so I am not sure. It's a style tradeoff. Anyhow I've reverted it since it seems confusing and might have some unforeseen consequences

**Comment by teto** [comment]
@MattSturgeon thanks for the [pkgs.testers.testBuildFailure](https://nixos.org/manual/nixpkgs/unstable/#tester-testBuildFailure) tip.
I've tried to use it but seems like it doesn't like the __structuredAttrs check because of shell hardening.
 ```
$ nix-build -A neovim.tests.failed_check
these 2 derivations will be built:
  /nix/store/1w8g4rcvzazrz1zlk657gclnpavn5h6q-init.lua.drv
  /nix/store/8wz4lh3a825mvfa40miadpzcnmlrarf5-neovim-0.10.2.drv
building '/nix/store/1w8g4rcvzazrz1zlk657gclnpavn5h6q-init.lua.drv'...
building '/nix/store/8wz4lh3a825mvfa40miadpzcnmlrarf5-neovim-0.10.2.drv'...
testBuildFailure: Expecting non-zero exit from builder and args: '/nix/store/mm0pa3z7kk6jh1i9rkxqxjqmd8h1qpxf-bash-5.2p37/bin/bash' '-e' '/nix/store/v6x3cs394jgqfbi0a42pam708flxaphh-default-builder.sh'
original builder: structuredAttrs is enabled
original builder: Running phase: patchPhase
original builder: @nix { "action": "setPhase", "phase": "patchPhase" }
original builder: Running phase: updateAutotoolsGnuConfigScriptsPhase
original builder: @nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
original builder: Running phase: configurePhase
original builder: @nix { "action": "setPhase", "phase": "configurePhase" }
....
       > original builder: _addToLuaPath called for dir /nix/store/j7065pwakq2f2nrshwi5c6qh7fnqg3d0-vim-pack-dir
       > original builder: LUA_PATH towards the end of packdir: ;;
       > original builder: Running phase: checkPhase
       > original builder: @nix { "action": "setPhase", "phase": "checkPhase" }
       > original builder: Error detected while processing /nix/store/bx0dbsfnnkg94yaarw9livpwxk49anvg-init.lua:
       > testBuildFailure: Original builder produced exit code: 1
       > /nix/store/ms79njxnpys354nc5n9v27qh4i5g1gjj-expect-failure.sh: line 37: __structuredAttrs: unbound variable
       For full logs, run 'nix log /nix/store/8wz4lh3a825mvfa40miadpzcnmlrarf5-neovim-0.10.2.drv'.
       
```
@Artturin seems like you touched this last. Is it because of `set -u` ? 
  should ` if [ -n "$__structuredAttrs" ]; then` be `if [ -v "__structuredAttrs" ]; then` instead ? I tried and it seems to work until I get to 
  ```
          echo "${!outputs[*]}"
```
which fails with 
```
       > original builder: @nix { "action": "setPhase", "phase": "checkPhase" }
       > original builder: Error detected while processing /nix/store/bx0dbsfnnkg94yaarw9livpwxk49anvg-init.lua:
       > testBuildFailure: Original builder produced exit code: 1
       > /nix/store/gb8s8251da5zap4d1rxlpiv37skim3gm-expect-failure.sh: line 40: outputs: unbound variable
```
Should I wrap the call in 
```
runCommand "example" {
  failed = testers.testBuildFailure (....
} ''
  grep -F 'failing though' $failed/testBuildFailure.log
  [[ 3 = $(cat $failed/testBuildFailure.exit) ]]
  touch $out
''
```
as presented in the doc ?

**Comment by MattSturgeon** [comment]
> @MattSturgeon thanks for the [pkgs.testers.testBuildFailure](https://nixos.org/manual/nixpkgs/unstable/#tester-testBuildFailure) tip.

No problem 😁 

> I've tried to use it but seems like it doesn't like the `__structuredAttrs` check because of shell hardening.

🤔


> Should I wrap the call in 
> ```
> runCommand "example" {
>   failed = testers.testBuildFailure (....
> } ''
>   grep -F 'failing though' $failed/testBuildFailure.log
>   [[ 3 = $(cat $failed/testBuildFailure.exit) ]]
>   touch $out
> ''
> ```
> as presented in the doc ?
> 

That's what we do in nixvim; pass the failing derivation into another test that can assert it failed in the expected way.

**Review by PerchunPak — APPROVED** [review]
If nixpkgs-review passes, LGTM. Nice and simple feature

Though maybe document it somewhere?

File: pkgs/applications/editors/neovim/wrapper.nix [hunk 3]
```diff
@@ -250,6 +250,13 @@ let
     # A Vim "package", see ':h packages'
     vimPackage = myVimPackage;
 
+    checkPhase = ''
+      runHook preCheck
+
+      $out/bin/nvim -i NONE -e +quitall!
+      runHook postCheck
+      '';
```
Comment (GaetanLepage) [inline pkgs/applications/editors/neovim/wrapper.nix:258]:
```suggestion
    '';
```

**Comment by teto** [comment]
> pass the failing derivation into another test that can assert it failed in the expected way.

it would be better but this should work as is, without needing to check for the failure.

> Though maybe document it somewhere?

Issue is this appears only in the unstable wrapper. Neovim being popular we've tried to maintain compatibility with the old wrapper to not break configs. I think the new wrapper is better enough than the old one so we can switch. It is something to approach carefully though so let it be for 25.04. 
I really want to have the doc generated from code as well so I've opened https://github.com/nix-community/nixdoc/issues/137 . 

I would like to use this PR feature to update the neovim tests so I am going to merge even if it means disabling the doCheck test. I've opened https://github.com/NixOS/nixpkgs/pull/355953 for a fix to enable the test.
